### PR TITLE
Implicit grouping by count throws 120

### PR DESCRIPTION
**Pull request by danielbeeke**
Hi Ruben and others,

I think I may have finished #120.

- Renamed skipUngroupedVariableCheck to skipValidation
- Added skipValidation to readme
- I did not find other places where adding skipValidation looked useful. If this is incorrect I can add it. 
- Added a new folder for testing queries that only work while skipping the validation
- Added one query that should have been added before. It was at the top of #120 
- Fixed the call to .map when there was no array
- Spec seems to work, tests also. It does however give notices about .tsv files.

### CODE REVIEW

**Comment by danielbeeke** [comment]
One small thing that I am wondering about.. the names of the queries do they have a relationship with the spec?
If so I might need to rename the newly added queries.

File: README.md [hunk 1]
```diff
@@ -28,6 +28,24 @@ parsedQuery.variables = ['?mickey'];
 var generatedQuery = generator.stringify(parsedQuery);
 ```
 Set `sparqlStar` to `true` to allow [SPARQL*](https://blog.liu.se/olafhartig/2019/01/10/position-statement-rdf-star-and-sparql-star/) syntax.
+
+
+### Validation
+
```
Comment (RubenVerborgh) [inline README.md:34]:
```suggestion

### Validation
```

File: README.md [hunk 2]
```diff
@@ -28,6 +28,24 @@ parsedQuery.variables = ['?mickey'];
...
+By default SPARQL.js throws on validation errors.
```
Comment (RubenVerborgh) [inline README.md:35]: Perhaps explain what they are; "query validation" as such is not a thing.

Maybe queries that are syntactically correct, but not allowed by the spec?

File: README.md [hunk 3]
```diff
@@ -28,6 +28,24 @@ parsedQuery.variables = ['?mickey'];
...
+
+```JavaScript
+// Parse a SPARQL query without validation.
+var SparqlParser = require('sparqljs').Parser;
+var parser = new SparqlParser({ skipValidation: true });
+var parsedQuery = parser.parse(
+  'select (?x as ?xString)' +
+  '(count(?y) as ?count)' +
+  '{ ?x ?y ?z }');
+
```
Comment (RubenVerborgh) [inline README.md:46]:
```suggestion
  '{ ?x ?y ?z }');
```

(remove the newline)

File: lib/sparql.jison [hunk 4]
```diff
@@ -642,7 +642,7 @@ SelectQuery
     | SelectClauseVars     DatasetClause* WhereClause SolutionModifier
     {
       // Check for projection of ungrouped variable
-      if (!Parser.skipUngroupedVariableCheck) {
+      if (!Parser.skipValidation) {
```
Comment (RubenVerborgh) [inline lib/sparql.jison:645]:
```suggestion
      if (!Parser.skipUngroupedVariableCheck && !Parser.skipValidation) {
```

so we maintain compatibility for the entire major version.

File: sparql.js [hunk 5]
```diff
@@ -13,7 +13,7 @@ module.exports = {
    *   sparqlStar?: boolean,
    * }
    */
-  Parser: function ({ prefixes, baseIRI, factory, sparqlStar, skipUngroupedVariableCheck } = {}) {
+  Parser: function ({ prefixes, baseIRI, factory, sparqlStar, skipValidation } = {}) {
```
Comment (RubenVerborgh) [inline sparql.js:16]: We need to keep it here too for backward compatibility.

**Comment by RubenVerborgh** [comment]
>  the names of the queries do they have a relationship with the spec?

Yes, queries like `sparql-10-2-2c.sparql` point to section 10.2.2 of the spec.

So probably best to rename indeed.

**Review by RubenVerborgh — CHANGES_REQUESTED** [review]
Thanks, @danielbeeke. Looks good, except that we need to maintain backwards compatibility, so the existing option flag still has to work. I'm fine with introducing another name as well; just both have to work.

**Comment by rubensworks** [comment]
Thanks @danielbeeke!

> Spec seems to work, tests also. It does however give notices about .tsv files.

You can ignore this, it's a warning that is unrelated to SPARQL.js.

File: lib/sparql.jison (hunk 4 above)
Comment (danielbeeke) [reply lib/sparql.jison:645]: I fixed it in sparql.js, that way the parser code looks cleaner to me. Would this be a good way or would you rather have it in the way you have suggested?

**Comment by RubenVerborgh** [comment]
All good, thanks!

**Pull request merged**